### ISSUE TITLE
Add out parameter

### ISSUE THREAD

**Issue by eyonland** [tenstorrent-metal/tt-metal#5044]

**Comment by eyonland** [comment]
MNIST
Linear
LogSoftmax
NLLLoss
SGD
Argmax
ElemwiseBinary/ELEMWISE_BINARY_EQ
Typecast
Sum
GPT2
BatchedMatmul
ElemwiseBinary/ELEMWISE_BINARY_ADD
ElemwiseBinary/ELEMWISE_BINARY_ADDALPHA
ElemwiseBinary/ELEMWISE_BINARY_MUL
ElemwiseUnary/ELEMWISE_UNARY_ADDElemwiseUnary/ELEMWISE_UNARY_ASSIGN
ElemwiseUnary/ELEMWISE_UNARY_DIV
ElemwiseUnary/ELEMWISE_UNARY_EXP
ElemwiseUnary/ELEMWISE_UNARY_FILL
ElemwiseUnary/ELEMWISE_UNARY_FILLZERO
ElemwiseUnary/ELEMWISE_UNARY_MUL
ElemwiseUnary/ELEMWISE_UNARY_POW
ElemwiseUnary/ELEMWISE_UNARY_SQRT
ElemwiseUnary/ELEMWISE_UNARY_TANH
Layernorm
Linear
Matmul
Softmax
AdamW
Arange
ClipGradNorm
ElemwiseTernary/ELEMWISE_TERNARY_ADDCDIV
ElemwiseTernary/ELEMWISE_TERNARY_ADDCMUL
Embedding
GenerateRandomBitMask
NLLLoss
NLLLossMeanPre
NLLLossUnreduced
Repeat
RepeatBackward
Sum
Typecast
Where
DropoutScaling
Groupnorm
MaskedFill
Prod
Tril

**Comment by jliangTT** [comment]
@razorback3 - FYI the feature for Output tensor as argument of the OP API

**Comment by jliangTT** [comment]
@muthutt , will you cover this function in #5163 as part of the migration? Are you still on-track?

**Comment by eyonland** [comment]
Would you rebase on main?
Seeing a lot files changed trying to compare with: [https://github.com/tenstorrent-metal/tt-metal/compare/seunghwan100/support_optional_output_tensors...main](https://github.com/tenstorrent-metal/tt-metal/compare/seunghwan100/support_optional_output_tensors...main)

**Comment by jliangTT** [comment]
As discussed in the slack - @razorback3 has agreed to split the roll out according to the ownership as follows (this is the same as spreadsheet) 

`MNIST
M - Linear - #5044: Add output tensors support for linear op #6259
M- LogSoftmax
M - NLLLoss
M - SGD
T - Argmax
T - ElemwiseBinary/ELEMWISE_BINARY_EQ
T - Typecast
M - Sum - Add optional output tensor for Moreh_sum #6591 
GPT2
M - BatchedMatmul - Add optional tensor support for bmm #6541
T - ElemwiseBinary/ELEMWISE_BINARY_ADD
T - ElemwiseBinary/ELEMWISE_BINARY_ADDALPHA
T - ElemwiseBinary/ELEMWISE_BINARY_MUL
T - ElemwiseUnary/ELEMWISE_UNARY_ADD
T - ElemwiseUnary/ELEMWISE_UNARY_ASSIGN
T - ElemwiseUnary/ELEMWISE_UNARY_DIV
T - ElemwiseUnary/ELEMWISE_UNARY_EXP
T - ElemwiseUnary/ELEMWISE_UNARY_FILL
T - ElemwiseUnary/ELEMWISE_UNARY_FILLZERO
T - ElemwiseUnary/ELEMWISE_UNARY_MUL
T - ElemwiseUnary/ELEMWISE_UNARY_POW
T - ElemwiseUnary/ELEMWISE_UNARY_SQRT
T - ElemwiseUnary/ELEMWISE_UNARY_TANH
M - Layernorm
M - Linear
M - Matmul - Add output tensors support for matmul #6588 (moreh)
M - Softmax
M - AdamW
M - Arange
M - ClipGradNorm
T - ElemwiseTernary/ELEMWISE_TERNARY_ADDCDIV
T -ElemwiseTernary/ELEMWISE_TERNARY_ADDCMUL
T - Embedding
M - GenerateRandomBitMask
M - NLLLoss
M - NLLLossMeanPre
M - NLLLossUnreduced
T - Repeat
T - RepeatBackward
M - Sum
T -Typecast
T - Where
M - DropoutScaling
M - Groupnorm
M - MaskedFill
T - Prod
M - Tril`

**Comment by razorback3** [comment]
@jliangTT 
As we split the ownership, is it okay to ignore issues/pull-requests related to this issue and raised by Metal team on Moreh ops?

**Comment by jliangTT** [comment]
yes, we should abandon those PRs. @eyonland @umadevimcw , are you in agreement with that?

**Comment by eyonland** [comment]
@razorback3 and @jliangTT, we should close these pull requests that have moreh ops on them.  

If I need to make an infra change, to demonstrate a proposal, I'll make a new PR and explicitly state that.

**Comment by jliangTT** [comment]
@eyonland , what is the latest status for this?

**Comment by KalaivaniMCW** [comment]
@eyonland Could you take a look at this [commit](https://github.com/tenstorrent-metal/tt-metal/commit/adfdba6c6942a6b98659c48a6dbd1339e82f5773) - (changes for adding optional output tensors to binary ops.) 
I am facing build issue 
![image (37)](https://github.com/tenstorrent-metal/tt-metal/assets/156762498/922e5ee2-5e46-4197-af45-e5dda0ea3ccc)

**Comment by jliangTT** [comment]
This is currently blocked from TTNN refactor items. Our review with @eyonland is we need to sync up internally to see when we can execute on this.

**Comment by davorchap** [comment]
The highest priority list:
```
Linear
Softmax
NLLLoss
SGD
Argmax
ElemwiseBinary/ELEMWISE_BINARY_EQ
Typecast
Sum
```

**Comment by KalaivaniMCW** [comment]
Hi
I have added optional output tensor to binary ops,  with operation::launch_op, operation::run  and removed autoformat operations. I have made fixes for those ops that were dependent on autoformat functionality. 
Except for prod_bw all the other ops support the changes (some issues with permute and unpad). 
CI : https://github.com/tenstorrent/tt-metal/actions/runs/9112666143/
draft PR : https://github.com/tenstorrent/tt-metal/pull/8394

**Comment by razorback3** [comment]
@dongjin-na Would you check the above PR?

**Comment by dongjin-na** [comment]
I reviewed the 8394 PR and ran test_eltwise_binary_optional_output.py. Through testing, I confirmed that this PR supports the optional output tensor for eltwise_binary_op. Although the test items were EQ, MUL, and ADD, I expect other ops will also function well.

Here are a few comments regarding this:
- Is there a reason to use std::vector<std::optional<Tensor>> instead of std::optional<Tensor> for opt_output_tensor?
- Will optional output tensor support for backward and composite ops, except prod_bw, be implemented after the 8394 PR?

**Comment by davorchap** [comment]
MNIST

* Linear - https://github.com/tenstorrent/tt-metal/pull/6259
* LogSoftmax
* NLLLoss
* SGD
* Argmax
* ElemwiseBinary/ELEMWISE_BINARY_EQ: #8540 
* Typecast: #4858 
* Sum - https://github.com/tenstorrent/tt-metal/issues/6591

**Comment by KalaivaniMCW** [comment]
Added optional output tensor for eltwise binary ops (fwd)
- ElemwiseBinary/ELEMWISE_BINARY_EQ: #8394 (merged to main)
CC: @davorchap @eyonland

**Comment by davorchap** [comment]
what's the status of these for optional output tensor @umadevimcw 

MNIST

Linear - https://github.com/tenstorrent/tt-metal/pull/6259
LogSoftmax
NLLLoss
SGD
Argmax
ElemwiseBinary/ELEMWISE_BINARY_EQ - https://github.com/tenstorrent/tt-metal/pull/8394
Typecast
Sum - https://github.com/tenstorrent/tt-metal/issues/6591
